### PR TITLE
Add chart_kit to core extensions directory

### PR DESCRIPTION
**Pull request by ufundo**
Overview
----------------------------------------
Adds the chart_kit extension (as developed at https://lab.civicrm.org/ufundo/chart_kit/ ) to the core extensions directory.

What is Chart Kit?
-------------------------------------------
Chart Kit is an extension which provides an additional display type to Search Kit called "Chart", which allows building a chart from the Search Kit results.

image

image

Under the hood, the search kit result data is passed to the DC charting library, which renders an SVG on the screen.

The search admin UI includes various config options for how the data is loaded ( chart type, which columns etc )

image

Before
----------------------------------------
- Chart Kit only installable from source

After
----------------------------------------
- extension is available for anyone to enable
- Chart Kit can start to be used in admin_ui / search_ui extensions to provide inline visualisations / user-configurable alternatives to existing core UI charts
- updates to search_kit and chart_kit interface can be easily co-ordinated
- sites on older core versions can still install the version of Chart Kit from Gitlab if desired as there is no change in the extension's fundamental compatibility requirements. but upgrading to 5.77 (?) will require co-ordinating removal of the extension from the site extension directory

Technical Details
----------------------------------------
- the library dependencies are added to the main `composer.json` file, but namespaced as the versions "owned" by the chart_kit extension, which live in the extension's `packages` directory
- the chart_kit extension provides its own "chart_kit" bundle, which is very similar to the core visual bundle, but using higher versions of the libraries

### CODE REVIEW

**Comment by colemanw** [comment]
> the library dependencies are added to the main composer.json file, but namespaced as the versions "owned" by the chart_kit extension

What purpose does the namespacing serve? Typically any library provided by core is available everywhere, not just to one extension.

**Comment by ufundo** [comment]
> What purpose does the namespacing serve? Typically any library provided by core is available everywhere, not just to one extension.

So my concern based on looking at the pre-existing dc / d3 copies is that if adding in globally, it's somewhat hard to upgrade the library version, because you don't know who / what might be being using it downstream, and expecting that version.

I figured the sensible options are:

**A) add the libraries named at a specific version (in this case add `dc-4.1.x` and `d3-6.7.x`)**
- this is good because anything else who wants to use them can, with a guarantee of version
- however if chart_kit wants to upgrade, it would involve adding another copy, and then not knowing whether you can remove the old one, because now you don't know who / what when might be relying on them (in the older version)

**B) add the libraries as "chart kit's versions"**  
- this doesn't stop anything else using them, but there's an understanding that they will be whatever versions chart kit wants them to be
- if it's time to upgrade the libraries for chart kit, then go for it - anyone else re-using these libraries will have to get with the programme... (or bring their own versions)

To me B felt like the more "extension first" approach - particularly with these kind of libraries where I think most uses are quite tightly coupled with the code. (As compared with say, the icon library, where it makes sense for icons to be consistent across all the different extensions on a site).

Open to A or other suggestions though :slightly_smiling_face:

**Comment by ufundo** [comment]
Tests pass :)

I've installed on the test site and made some example contribution graphs, and added to the dashboard for the admin user.

I found two new bugs :grin: ....

The (new) date rounding rounds UP, so it lists contributions from half way through 2024 as "2025". Will fix.

The charts are currently a fixed pixel width. This can be good for making sure things fit. But it does not work well for contribution dashboard widgets. I'm not sure there's an easy fix for that in the short term.

File: ext/chart_kit/info.xml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  <name>Chart Kit</name>
+  <description>This extension provides additional Search Display types for Search Kit for making charts, using the dc/d3 library.</description>
+  <license>AGPL-3.0</license>
+  <maintainer>
+    <author>Benjamin Walpole</author>
+    <email>ben@civicrm.org</email>
+  </maintainer>
+  <urls>
+    <url desc="Documentation">https://docs.civicrm.org/user/en/latest/the-user-interface/search-kit/</url>
+    <url desc="Chat">https://chat.civicrm.org/civicrm/channels/search-improvements</url>
```
Comment (ufundo) [inline ext/chart_kit/info.xml:13]: wondering about whether to create another MM channel separate from the main Search Kit one.

@colemanw do you have a view?

Comment (ufundo) [reply ext/chart_kit/info.xml:13]: I made https://chat.civicrm.org/civicrm/channels/dataviz

**Comment by eileenmcnaughton** [comment]
merging with chart-kit as is based on discussion on CT channel that this is at least OK for now

**Comment by totten** [comment]
I'm OK with merging with an extra copy of d3/dc/crossfilter.

* There is a technical-debt here in having multiple copies/old copies. However, organizationally, it's the same debt for us regardless of which repo `chart_kit` is tracked in. It's more a question of prioritization.
* The other references could become outmoded at higher level.
* If we do wind up with these multiple copies long-term (eg because `d3@3.x` and `d3@6.x` are too difficult to migrate), then I'd be tempted to reshuffle the naming. (`visual-bundle@3.x` and `visual-bundle@6.x`.) But I'd let that idea marinate longer.

**Comment by ufundo** [comment]
:bar_chart: :tada: !

I have tried to tackle the debt related to the old versions here: https://github.com/civicrm/civicrm-core/pull/30796 But stuck on a mysterious test error due to `d3` reference in the karma tests that I can't pin down.

> If we do wind up with these multiple copies long-term (eg because d3@3.x and d3@6.x are too difficult to migrate), then I'd be tempted to reshuffle the naming. (visual-bundle@3.x and visual-bundle@6.x.) But I'd let that idea marinate longer.

That makes sense to me - but a) where should these bundles be registered? b) it feels less "LEAP"y - you end up with coordination problems about who is using the visual-bundle at this or that version and whether it can be removed now. It seems unlikely to me that any given site will have more than 1 or 2 different (extension) consumers of these libraries, so it's reasonable they each just have their own copy with whatever version they need.

**Pull request merged**